version: '3'

services:

  db:
    image: postgres
    container_name: db
    environment:
      - POSTGRES_PASSWORD=mysecretpassword
      - PGDATA=/var/lib/postgresql/data/pgdata
      - POSTGRES_DB=postgres_docker
    volumes:
      - pg_data:/var/lib/postgresql/data/pgdata
    env_file:
      - .env
    ports:
      - '54321:54321'
    healthcheck:
      test: [ "CMD-SHELL", "pg_isready -U postgres" ]
      interval: 10s
      timeout: 5s
      retries: 5
#
  redis:
    image: redis
    container_name: redis_cont
    ports:
      - '6379:6379'
    volumes:
      - redis_data:/data
    healthcheck:
      test: [ "CMD", "redis-cli", "ping" ]
      interval: 10s
      timeout: 5s
      retries: 3

  app:
    build: .
    container_name: app_cont
    tty: true
    command: sh -c "python3 manage.py migrate && python3 manage.py runserver 0.0.0.0:8000"
    ports:
      - '8000:8000'
    depends_on:
      db:
        condition: service_healthy

  celery:
    build: .
    container_name: celery_cont
    tty: true
    command: celery -A config worker -l INFO
    depends_on:
      redis:
        condition: service_healthy
    environment:
      - CELERY_BROKER_URL=redis://redis:6379/0
      - CELERY_LOG_LEVEL=debug
    healthcheck:
      test: [ "CMD", "celery", "inspect", "ping" ]
      interval: 10s
      timeout: 5s
      retries: 3

  celery-beat:
    build: .
    container_name: celery_beat_cont
    command: celery -A config beat -l info -S django --loglevel warning
    volumes:
      - celery_beat_data:/app
    depends_on:
      redis:
        condition: service_healthy
      celery:
        condition: service_healthy
    environment:
      - CELERY_BROKER_URL=redis://redis:6379/0
      - CELERY_LOG_LEVEL=debug

volumes:
  pg_data:
  redis_data:
  celery_data:
  celery_beat_data:
